crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs

```diff
@@ -0,0 +1,261 @@
+                            .context("failed to protect stack guard page")?;
+                    }
+                }
+            }
+
+            Ok(Self {
+                mapping,
+                stack_size,
+                max_stacks,
+                page_size,
+                async_stack_zeroing: false,
+                async_stack_keep_resident: 0,
+                index_allocator: SimpleIndexAllocator::new(total_stacks),
+            })
+        }
+
+        /// Allocate a new fiber.
+        pub fn allocate(&self) -> Result<FiberStack> {
+            if self.stack_size == 0 {
+                bail!("pooling allocator not configured to enable fiber stack allocation");
+            }
+
+            let index = self
+                .index_allocator
+                .alloc()
+                .ok_or_else(|| {
+                    anyhow!(
+                        "maximum concurrent fiber limit of {} reached",
+                        self.max_stacks
+                    )
+                })?
+                .index();
+
+            assert!(index < self.max_stacks);
+
+            unsafe {
+                // Remove the guard page from the size
+                let size_without_guard = self.stack_size - self.page_size;
+
+                let bottom_of_stack = self
+                    .mapping
+                    .as_ptr()
+                    .add((index * self.stack_size) + self.page_size)
+                    .cast_mut();
+
+                commit_pages(bottom_of_stack, size_without_guard)?;
+
+                let stack = super::FiberStack::from_raw_parts(bottom_of_stack, size_without_guard)?;
+                Ok(stack)
+            }
+        }
+
+        /// Deallocate a previously-allocated fiber.
+        ///
+        /// # Safety
+        ///
+        /// The fiber must have been allocated by this pool, must be in an allocated
+        /// state, and must never be used again.
+        pub unsafe fn deallocate(&self, stack: &FiberStack) {
+            let top = stack
+                .top()
+                .expect("fiber stack not allocated from the pool") as usize;
+
+            let base = self.mapping.as_ptr() as usize;
+            let len = self.mapping.len();
+            assert!(
+                top > base && top <= (base + len),
+                "fiber stack top pointer not in range"
+            );
+
+            // Remove the guard page from the size
+            let stack_size = self.stack_size - self.page_size;
+            let bottom_of_stack = top - stack_size;
+            let start_of_stack = bottom_of_stack - self.page_size;
+            assert!(start_of_stack >= base && start_of_stack < (base + len));
+            assert!((start_of_stack - base) % self.stack_size == 0);
+
+            let index = (start_of_stack - base) / self.stack_size;
+            assert!(index < self.max_stacks);
+
+            if self.async_stack_zeroing {
```

Comment:
Hm, it seems a bit weird to re-use `async`s config options here. How about dropping the zero-ing support for now until we add our own corresponding flags?

Reply:
Can do. I plan to experiment with options like these soon.